README.md
# EnhancedEndpointTracker

The EnhancedEndpointTracker is a Cisco ACI application that maintains a database of endpoint
events on a per-node basis allowing for unique fabric-wide analysis. The application can be
configured to analyze, notify, and automatically remediate various endpoint events. This gives
ACI fabric operators better visibility and control over the endpoints in the fabric.

Features include:

- Easy to use GUI with fast type-ahead search for any mac/ipv4/ipv6 endpoint in the fabric
- Move, offsubnet, rapid, and stale endpoint analysis
- Configurable notification options via syslog/email for various events detected by the app
- Capability to clear an endpoint on one or more nodes via GUI or API
- Distributed architecture allowing app to grow based on the size of the fabric
- Fully documented swagger API to allow for easy integration with other tools.

This application is written in python and utilizes a distributed MongoDB database to persist data.
It also relies on a custom Flask framework for handling API requests, Redis for messaging between
components, and Angular with CiscoUI for the frontend.

For more details, refer to the online documentation at:
http://aci-enhancedendpointtracker.readthedocs.io/en/latest/ (http://aci-enhancedendpointtracker.readthedocs.io/en/latest/)


docs/releases.rst
.. _releases:

Releases
========

This page will track information about each new release along with new features and any known
issues. 

Version 2.1.2a
--------------
*Released Sep 12 2019*

* fix for Docker swarm node IP discovery
* change default worker count from 10 to 23 for swarm
* ensure web service is closing redis connections for each pub/sub message
* ensure all messages are sent on support qnum (regression from queue count reduction fixes)
* force username for ssh remote login procedures to ensure custom domain is maintained

Version 2.1.2
-------------
*Released Aug 20 2019*

* fix for #46 fabricNode parsing broken on recent build breaks all eptNode tables
* update worker hash calculation for better load-balancing of work in scale setups
* use pub/sub redis messaging for broadcast between subscriber and workers
* reduce queue count and queue stat objects when using pub/sub broadcasts
* address timeouts on techsupport collection in multi-node standalone cluster
* reduce memory utilization on startup by streaming class queries
* include sort on all concrete objects initializations to prevent out-of-order results during paging
* use local user directory instead of global tmp for compose and deploy logs to prevent access
  problems for multiuser deployments
* additional user validation checks during docker swarm deployment
* increase manager timeout on app-status to support scale setups
* fix mongo cursor timeouts due to docker routed-mesh mongos load-balancing in multi-node standalone
  cluster
* propagate session/subscriber errors into fabric events so user can quickly isolate subscription
  restart errors
* added UI and backend support for configurable heartbeat interval, timeout, and retry count
* improve queue cleanup on fabric restart
* logrotate updates for apache logs
* UI polling service fix to prevent rapid requests on 400/500 errors
* UI optimize fabric status and manager polling mechanisms
* UI fix for cascading polling events on refresh

Version 2.1.1
-------------
*Released Mar 21 2019*

* Configurable SMTP settings including custom SMTP port, TLS encryption, and support for SMTP relay
  with optional authentication
* Configurable session timeout
* Configurable Refresh time for subscriptions. Staring in ACI 4.0 the subscription refresh time can 
  be extended up to 10 hours but will be limited to 60 seconds in older versions. The app will 
  auto-detect the version of code and apply a max of 60 seconds if APIC or switch is <4.0.
  Previously this was static at 30 seconds.
* Moved managed object event handling from subscriber process to dedicated worker. This addresses 
  issues such as high rate of tunnelIf events during initial build that can cause subscriber process 
  to hang.
* Cross reference fabricNode and topSystem to ensure inactive nodes are included in initial build. 
  This resolves a bug TEP calculation for vpcs if one node in the vpc is offline when app starts
* Disable accurate queue-length calculation on UI as it impacts manager process under scale
* Updates to app deployment scripts to allow user to pass in all arguments via command-line in 
  addition to editing cluster yaml file
* Trigger fabric restart on worker heartbeat failure and new worker detection
* Removed exim4 dependencies and rely on native python smtplib for email notifications


Version 2.0.x
-------------
*Released Feb 22 2019*

Initial 2.0 release. This was a complete rewrite with focus on scalability while maintaining and 
augmenting the feature set from the 1.x versions. Features include:

* Support for full endpoint scale fabric
* Easy to use GUI with fast type-ahead search for any mac/ipv4/ipv6 endpoint in the fabric
* Full details on current state of an endpoint in the fabric along with historical information
* Move, offsubnet, rapid, and stale endpoint analysis
* Configurable notification options via syslog/email for various events detected by the app
* Capability to clear an endpoint on one or more nodes via GUI or API
* Distributed architecture allowing app to grow based on the size of the fabric
* Fully documented swagger API to allow for easy integration with other tools
* Flexible deployment options include APIC and APIC-mini apps along with standalone all-in-one
  container and standalone docker-swarm cluster.




app/views/auth.py

from flask import current_app, Blueprint, render_template
auth_prefix = "/auth"
auth = Blueprint("auth", __name__, url_prefix=auth_prefix)

from flask import Flask, jsonify, flash, redirect, url_for
from flask import request, make_response, g, url_for, abort, session
from flask_login import (LoginManager, login_required, login_user,
    current_user, logout_user)

from ..models.users import Users
from ..models.roles import Roles
from ..models.utils import MSG_403

# setup login manager
login_manager = LoginManager()

# since this is a blueprint, use record_once instead of login_manager.init_app
@auth.record_once
def on_load(state):
    # setup login manager login_view
    if state.app.config.get("SSO_ENABLED", False):
        # url_for not available at this point 
        login_manager.login_view = "%s/sso/" % auth_prefix
    else:
        # url_for not available at this point 
        login_manager.login_view = "%s/login/" % auth_prefix
    login_manager.login_message = ""
    login_manager.init_app(state.app)

@auth.before_app_request
def before_request():
    # force everything over HTTPS if enabled
    if current_app.config.get("force_https", False):
        fwd_proto = request.headers.get("x-forwarded-proto",None)
        if fwd_proto is not None:
            if fwd_proto.lower() == "http":
                return redirect(request.url.replace("http:","https:", 1))
        else:
            if re.search("^http:", request.url) is not None:
                return redirect(request.url.replace("http:","https:", 1))

    # set global object various configs
    g.app_name = current_app.config.get("app_name", "AppName1")

    # set global object 'g.user' based off current user session
    g.ROLE_FULL_ADMIN = Roles.FULL_ADMIN
    g.user = current_user
    if g.user is not None:
        if hasattr(g.user, 'role') and g.user.role == Roles.BLACKLIST:
            Users.logout()
            abort(403, MSG_403)
        elif not current_app.config.get("LOGIN_ENABLED", True) and \
            not g.user.is_authenticated:
            # auto-login user as local if login is disabled
            g.user = Users.load_user("local")
            if g.user is None:
                # setup local user
                from ..models.users import setup_local
                setup_local(current_app)
                g.user = Users({"username":"local", "role": Roles.FULL_ADMIN})
            Users.start_session("local")

@login_manager.user_loader
def load_user(username):
    return Users.load_user(username)

@auth.route("/login", methods=["GET", "POST"])
@auth.route("/login/", methods=["GET", "POST"])
def login():
    if request.method == "POST":
        if Users.login(
            username = request.form['username'],
            password = request.form['password'],
            remember = True):
            return redirect(request.args.get("next") or "/")
        else:
            flash("The email or password you entered is incorrect.")
            return render_template("auth/login.html")

    return render_template("auth/login.html")

@auth.route("/logout")
@auth.route("/logout/")
def logout():
    Users.logout()
    return render_template("auth/logout.html")

@auth.route("/pwreset/<string:key>", methods=["GET"])
def pwreset(key):
    return render_template("auth/pwreset.html", key=key)

##############################################################################
# Auth API, imported by api module
##############################################################################

def api_login():
    """ login to application and create session cookie.  Note, if sso
        authentication is enabled, then this function is not required. Simply
        provided the sso cookie and the application will authenticate it.
        Args:
            username(str): application username
            password(str): application password
        Returns:
            success(bool): successfully login
    """
    data = request.json
    if not data: abort(400, "Invalid JSON provided")
    if "username" not in data:
        abort(400, "Required parameter \"username\" not provided")
    if "password" not in data:
        abort(400, "Required parameter \"password\" not provided")

    if Users.login(
        username = data["username"],
        password = data["password"],
        remember = True):
        return jsonify({"success": True})
    else:
        abort(401, "Authentication Failed")

def api_logout():
    """ logout of application and delete session """
    Users.logout()
    return jsonify({"success": True})


##############################################################################
# 
#   Cisco SSO handlers
#
##############################################################################

import re, requests

@auth.route("/sso", methods=["GET"])
@auth.route("/sso/", methods=["GET"])
def sso_login():
    """
    use sso_handler to check if user is already authenticated.  If not,
    redirect user to sso_url with referer set to /rsso (return sso) and next
    argument set to original provided parameter.  If user is authenticated,
    login user locally to site and redirect to page at next argument
    """
    if not current_app.config.get("SSO_ENABLED", False):
        abort(400, "SSO authentication not enabled")

    if len(current_app.config.get("sso_url",""))<=0:
        abort(500, "sso_url not defined in app.config")
    if len(current_app.config.get("fqdn",""))<=0:
        abort(500, "fqdn not defined in app.config")

    # not implemented, force fail
    return render_template("auth/sso_failed.html")


@auth.route("/rsso", methods=["GET"])
@auth.route("/rsso/", methods=["GET"])
def sso_login_return():
    """
    this page is redirected to after user has completed authentication on
    SSO site. If username is valid, redirect to 'next' parameter, else display
    sso_failed page
    """
    if not current_app.config.get("SSO_ENABLED", False):
        abort(400, "SSO authentication not enabled")

    # not implemented, force fail
    return render_template("auth/sso_failed.html")


